File: core/src/main/scala/kafka/log/LogSegments.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package kafka.log

import java.io.File
import java.util.Map
import java.util.concurrent.{ConcurrentNavigableMap, ConcurrentSkipListMap}

import kafka.utils.threadsafe
import org.apache.kafka.common.TopicPartition

import scala.jdk.CollectionConverters._

/**
 * This class encapsulates a thread-safe navigable map of LogSegment instances and provides the
 * required read and write behavior on the map.
 *
 * @param topicPartition the TopicPartition associated with the segments
 *                        (useful for logging purposes)
 */
class LogSegments(topicPartition: TopicPartition) {

  /* the segments of the log with key being LogSegment base offset and value being a LogSegment */
  private val segments: ConcurrentNavigableMap[Long, LogSegment] = new ConcurrentSkipListMap[Long, LogSegment]

  /**
   * @return true if the segments are empty, false otherwise.
   */
  @threadsafe
  def isEmpty: Boolean = segments.isEmpty

  /**
   * @return true if the segments are non-empty, false otherwise.
   */
  @threadsafe
  def nonEmpty: Boolean = !isEmpty

  /**
   * Add the given segment, or replace an existing entry.
   *
   * @param segment the segment to add
   */
  @threadsafe
  def add(segment: LogSegment): LogSegment = this.segments.put(segment.baseOffset, segment)

  /**
   * Remove the segment at the provided offset.
   *
   * @param offset the offset to be removed
   */
  @threadsafe
  def remove(offset: Long): Unit = segments.remove(offset)

  /**
   * Clears all entries.
   */
  @threadsafe
  def clear(): Unit = segments.clear()

  /**
   * Close all segments.
   */
  def close(): Unit = values.foreach(_.close())

  /**
   * Close the handlers for all segments.
   */
  def closeHandlers(): Unit = values.foreach(_.closeHandlers())

  /**
   * Update the directory reference for the log and indices of all segments.
   *
   * @param dir the renamed directory
   */
  def updateParentDir(dir: File): Unit = values.foreach(_.updateParentDir(dir))

  /**
   * Take care! this is an O(n) operation, where n is the number of segments.
   *
   * @return The number of segments.
   *
   */
  @threadsafe
  def numberOfSegments: Int = segments.size

  /**
   * @return the base offsets of all segments
   */
  def baseOffsets: Iterable[Long] = segments.values().asScala.map(_.baseOffset)

  /**
   * @param offset the segment to be checked
   * @return true if a segment exists at the provided offset, false otherwise.
   */
  @threadsafe
  def contains(offset: Long): Boolean = segments.containsKey(offset)

  /**
   * Retrieves a segment at the specified offset.
   *
   * @param offset the segment to be retrieved
   *
   * @return the segment if it exists, otherwise None.
   */
  @threadsafe
  def get(offset: Long): Option[LogSegment] = Option(segments.get(offset))

  /**
   * @return an iterator to the log segments ordered from oldest to newest.
   */
  def values: Iterable[LogSegment] = segments.values.asScala

  /**
   * @return An iterator to all segments beginning with the segment that includes "from" and ending
   *         with the segment that includes up to "to-1" or the end of the log (if to > end of log).
   */
  def values(from: Long, to: Long): Iterable[LogSegment] = {
    if (from == to) {
      // Handle non-segment-aligned empty sets
      List.empty[LogSegment]
    } else if (to < from) {
      throw new IllegalArgumentException(s"Invalid log segment range: requested segments in $topicPartition " +
        s"from offset $from which is greater than limit offset $to")
    } else {
      val view = Option(segments.floorKey(from)).map { floor =>
        segments.subMap(floor, to)
      }.getOrElse(segments.headMap(to))
      view.values.asScala
    }
  }

  def nonActiveLogSegmentsFrom(from: Long): Iterable[LogSegment] = {
    val activeSegment = lastSegment.get
    if (from > activeSegment.baseOffset)
      Seq.empty
    else
      values(from, activeSegment.baseOffset)
  }

  /**
   * @return the entry associated with the greatest offset less than or equal to the given offset,
   *         if it exists.
   */
  @threadsafe
  private def floorEntry(offset: Long): Option[Map.Entry[Long, LogSegment]] = Option(segments.floorEntry(offset))

  /**
   * @return the log segment with the greatest offset less than or equal to the given offset,
   *         if it exists.
   */
  @threadsafe
  def floorSegment(offset: Long): Option[LogSegment] = floorEntry(offset).map(_.getValue)

  /**
   * @return the entry associated with the greatest offset strictly less than the given offset,
   *         if it exists.
   */
  @threadsafe
  private def lowerEntry(offset: Long): Option[Map.Entry[Long, LogSegment]] = Option(segments.lowerEntry(offset))

  /**
   * @return the log segment with the greatest offset strictly less than the given offset,
   *         if it exists.
   */
  @threadsafe
  def lowerSegment(offset: Long): Option[LogSegment] = lowerEntry(offset).map(_.getValue)

  /**
   * @return the entry associated with the smallest offset strictly greater than the given offset,
   *         if it exists.
   */
  @threadsafe
  def higherEntry(offset: Long): Option[Map.Entry[Long, LogSegment]] = Option(segments.higherEntry(offset))

  /**
   * @return the log segment with the smallest offset strictly greater than the given offset,
   *         if it exists.
   */
  @threadsafe
  def higherSegment(offset: Long): Option[LogSegment]  = higherEntry(offset).map(_.getValue)

  /**
   * @return the entry associated with the smallest offset, if it exists.
   */
  @threadsafe
  def firstEntry: Option[Map.Entry[Long, LogSegment]] = Option(segments.firstEntry)

  /**
   * @return the log segment associated with the smallest offset, if it exists.
   */
  @threadsafe
  def firstSegment: Option[LogSegment] = firstEntry.map(_.getValue)

  /**
   * @return the base offset of the log segment associated with the smallest offset, if it exists
   */
  private[log] def firstSegmentBaseOffset: Option[Long] = firstSegment.map(_.baseOffset)

  /**
   * @return the entry associated with the greatest offset, if it exists.
   */
  @threadsafe
  def lastEntry: Option[Map.Entry[Long, LogSegment]] = Option(segments.lastEntry)

  /**
   * @return the log segment with the greatest offset, if it exists.
   */
  @threadsafe
  def lastSegment: Option[LogSegment] = lastEntry.map(_.getValue)

  /**
   * @return an iterable with log segments ordered from lowest base offset to highest,
   *         each segment returned has a base offset strictly greater than the provided baseOffset.
   */
  def higherSegments(baseOffset: Long): Iterable[LogSegment] = {
    val view =
      Option(segments.higherKey(baseOffset)).map {
        higherOffset => segments.tailMap(higherOffset, true)
      }.getOrElse(collection.immutable.Map[Long, LogSegment]().asJava)
    view.values.asScala
  }

  /**
   * The active segment that is currently taking appends
   */
  def activeSegment = lastSegment.get

  def sizeInBytes: Long = LogSegments.sizeInBytes(values)

  /**
   * Returns an Iterable containing segments matching the provided predicate.
   *
   * @param predicate the predicate to be used for filtering segments.
   */
  def filter(predicate: LogSegment => Boolean): Iterable[LogSegment] = values.filter(predicate)
}

object LogSegments {
  /**
   * Calculate a log's size (in bytes) from the provided log segments.
   *
   * @param segments The log segments to calculate the size of
   * @return Sum of the log segments' sizes (in bytes)
   */
  def sizeInBytes(segments: Iterable[LogSegment]): Long =
    segments.map(_.size.toLong).sum

  def getFirstBatchTimestampForSegments(segments: Iterable[LogSegment]): Iterable[Long] = {
    segments.map {
      segment =>
        segment.getFirstBatchTimestamp()
    }
  }
}

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegments.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.storage.internals.log;

import org.apache.kafka.common.TopicPartition;

import java.io.File;
import java.io.IOException;
import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.OptionalLong;
import java.util.concurrent.ConcurrentNavigableMap;
import java.util.concurrent.ConcurrentSkipListMap;
import java.util.function.Predicate;
import java.util.stream.Collectors;

/**
 * This class encapsulates a thread-safe navigable map of LogSegment instances and provides the
 * required read and write behavior on the map.
 */
public class LogSegments {

    private final TopicPartition topicPartition;
    /* the segments of the log with key being LogSegment base offset and value being a LogSegment */
    private final ConcurrentNavigableMap<Long, LogSegment> segments = new ConcurrentSkipListMap<>();

    /**
     * Create new instance.
     *
     * @param topicPartition the TopicPartition associated with the segments
     *                        (useful for logging purposes)
     */
    public LogSegments(TopicPartition topicPartition) {
        this.topicPartition = topicPartition;
    }

    /**
     * Return true if the segments are empty, false otherwise.
     *
     * This method is thread-safe.
     */
    public boolean isEmpty() {
        return segments.isEmpty();
    }

    /**
     * Return true if the segments are non-empty, false otherwise.
     *
     * This method is thread-safe.
     */
    public boolean nonEmpty() {
        return !isEmpty();
    }

    /**
     * Add the given segment, or replace an existing entry.
     *
     * This method is thread-safe.
     *
     * @param segment the segment to add
     */
    public LogSegment add(LogSegment segment) {
        return this.segments.put(segment.baseOffset(), segment);
    }

    /**
     * Remove the segment at the provided offset.
     *
     * This method is thread-safe.
     *
     * @param offset the offset to be removed
     */
    public void remove(long offset) {
        segments.remove(offset);
    }

    /**
     * Clears all entries.
     *
     * This method is thread-safe.
     */
    public void clear() {
        segments.clear();
    }

    /**
     * Close all segments.
     */
    public void close() throws IOException {
        for (LogSegment s : values())
            s.close();
    }

    /**
     * Close the handlers for all segments.
     */
    public void closeHandlers() {
        for (LogSegment s : values())
            s.closeHandlers();
    }

    /**
     * Update the directory reference for the log and indices of all segments.
     *
     * @param dir the renamed directory
     */
    public void updateParentDir(File dir) {
        for (LogSegment s : values())
            s.updateParentDir(dir);
    }

    /**
     * Take care! this is an O(n) operation, where n is the number of segments.
     *
     * This method is thread-safe.
     *
     * @return The number of segments.
     *
     */
    public int numberOfSegments() {
        return segments.size();
    }

    /**
     * @return the base offsets of all segments
     */
    public Collection<Long> baseOffsets() {
        return values().stream().map(s -> s.baseOffset()).collect(Collectors.toList());
    }

    /**
     * Return true if a segment exists at the provided offset, false otherwise.
     *
     * This method is thread-safe.
     *
     * @param offset the segment to be checked
     */
    public boolean contains(long offset) {
        return segments.containsKey(offset);
    }

    /**
     * Retrieves a segment at the specified offset.
     *
     * This method is thread-safe.
     *
     * @param offset the segment to be retrieved
     *
     * @return the segment if it exists, otherwise None.
     */
    public Optional<LogSegment> get(long offset) {
        return Optional.ofNullable(segments.get(offset));
    }

    /**
     * @return an iterator to the log segments ordered from oldest to newest.
     */
    public Collection<LogSegment> values() {
        return segments.values();
    }

    /**
     * @return An iterator to all segments beginning with the segment that includes "from" and ending
     *         with the segment that includes up to "to-1" or the end of the log (if to > end of log).
     */
    public Collection<LogSegment> values(long from, long to) {
        if (from == to) {
            // Handle non-segment-aligned empty sets
            return Collections.emptyList();
        } else if (to < from) {
            throw new IllegalArgumentException("Invalid log segment range: requested segments in " + topicPartition +
                    " from offset " + from + " which is greater than limit offset " + to);
        } else {
            Long floor = segments.floorKey(from);
            if (floor != null)
                return segments.subMap(floor, to).values();
            return segments.headMap(to).values();
        }
    }

    public Collection<LogSegment> nonActiveLogSegmentsFrom(long from) {
        LogSegment activeSegment = lastSegment().get();
        if (from > activeSegment.baseOffset())
            return Collections.emptyList();
        else
            return values(from, activeSegment.baseOffset());
    }

    /**
     * Return the entry associated with the greatest offset less than or equal to the given offset,
     * if it exists.
     *
     * This method is thread-safe.
     */
    private Optional<Map.Entry<Long, LogSegment>> floorEntry(long offset) {
        return Optional.ofNullable(segments.floorEntry(offset));
    }

    /**
     * Return the log segment with the greatest offset less than or equal to the given offset,
     * if it exists.
     *
     * This method is thread-safe.
     */
    public Optional<LogSegment> floorSegment(long offset) {
        return floorEntry(offset).map(e -> e.getValue());
    }

    /**
     * Return the entry associated with the greatest offset strictly less than the given offset,
     * if it exists.
     *
     * This method is thread-safe.
     */
    private Optional<Map.Entry<Long, LogSegment>> lowerEntry(long offset) {
        return Optional.ofNullable(segments.lowerEntry(offset));
    }

    /**
     * Return the log segment with the greatest offset strictly less than the given offset,
     * if it exists.
     *
     * This method is thread-safe.
     */
    public Optional<LogSegment> lowerSegment(long offset) {
        return lowerEntry(offset).map(e -> e.getValue());
    }

    /**
     * Return the entry associated with the smallest offset strictly greater than the given offset,
     * if it exists.
     *
     * This method is thread-safe.
     */
    public Optional<Map.Entry<Long, LogSegment>> higherEntry(long offset) {
        return Optional.ofNullable(segments.higherEntry(offset));
    }

    /**
     * Return the log segment with the smallest offset strictly greater than the given offset,
     * if it exists.
     *
     * This method is thread-safe.
     */
    public Optional<LogSegment> higherSegment(long offset) {
        return higherEntry(offset).map(e -> e.getValue());
    }

    /**
     * Return the entry associated with the smallest offset, if it exists.
     *
     * This method is thread-safe.
     */
    public Optional<Map.Entry<Long, LogSegment>> firstEntry() {
        return Optional.ofNullable(segments.firstEntry());
    }

    /**
     * Return the log segment associated with the smallest offset, if it exists.
     *
     * This method is thread-safe.
     */
    public Optional<LogSegment> firstSegment() {
        return firstEntry().map(s -> s.getValue());
    }

    /**
     * @return the base offset of the log segment associated with the smallest offset, if it exists
     */
    public OptionalLong firstSegmentBaseOffset() {
        Optional<LogSegment> first = firstSegment();
        if (first.isPresent())
            return OptionalLong.of(first.get().baseOffset());
        return OptionalLong.empty();
    }

    /**
     * Return the entry associated with the greatest offset, if it exists.
     *
     * This method is thread-safe.
     */
    public Optional<Map.Entry<Long, LogSegment>> lastEntry() {
        return Optional.ofNullable(segments.lastEntry());
    }

    /**
     * Return the log segment with the greatest offset, if it exists.
     *
     * This method is thread-safe.
     */
    public Optional<LogSegment> lastSegment() {
        return lastEntry().map(e -> e.getValue());
    }

    /**
     * @return an iterable with log segments ordered from lowest base offset to highest,
     *         each segment returned has a base offset strictly greater than the provided baseOffset.
     */
    public Collection<LogSegment> higherSegments(long baseOffset) {
        Long higherOffset = segments.higherKey(baseOffset);
        if (higherOffset != null)
            return segments.tailMap(higherOffset, true).values();
        return Collections.emptyList();
    }

    /**
     * The active segment that is currently taking appends
     */
    public LogSegment activeSegment() {
        return lastSegment().get();
    }

    public long sizeInBytes() {
        return LogSegments.sizeInBytes(values());
    }

    /**
     * Returns an Iterable containing segments matching the provided predicate.
     *
     * @param predicate the predicate to be used for filtering segments.
     */
    public Collection<LogSegment> filter(Predicate<LogSegment> predicate) {
        return values().stream().filter(predicate).collect(Collectors.toList());
    }

    /**
     * Calculate a log's size (in bytes) from the provided log segments.
     *
     * @param segments The log segments to calculate the size of
     * @return Sum of the log segments' sizes (in bytes)
     */
    public static long sizeInBytes(Collection<LogSegment> segments) {
        return segments.stream().mapToLong(s -> s.size()).sum();
    }

    public static Collection<Long> getFirstBatchTimestampForSegments(Collection<LogSegment> segments) {
        return segments.stream().map(s -> s.getFirstBatchTimestamp()).collect(Collectors.toList());
    }
}
